resolversrv/tests/unittests/test_db_query.py

```diff
@@ -8,186 +8,424 @@
+                                                        u'title': u'Part 13'
+                                                  }],
+                                     u'link_type': u'ASSOCIATED'}}
+                         )
 
 
-    # return 404 for not finding any records
     def test_link_associated_error_bibcode(self):
+        """
+        return 404 for not finding any records
+        :return:
+        """
         results = get_records(bibcode='errorbibcode', link_type='ASSOCIATED')
         response = LinkRequest(bibcode='errorbibcode', link_type='ASSOCIATED').request_link_type_associated(results)
         self.assertEqual(response._status_code, 404)
 
 
-    # returning list of urls
-    def test_link_article(self):
+    def test_link_associated_KeyError(self):
+        """
+        return 400 from request_link_type_associated where there is KeyError
+        :return:
+        """
+        results = [{'bibcode': u'1971ATsir.615....4D',
+                    'link_type': u'ASSOCIATED',
+                    'link_sub_type': '',
+                    'title': [u'Part 11', u'Part 10', u'Part 13', u'Part 12', u'Part  8', u'Part  9', u'Part  2', u'Part  3', u'Part  1', u'Part  6', u'Part  7', u'Part  4', u'Part  5'],
+                    'url_KeyError': [u'1971ATsir.615....4D', u'1971ATsir.624....1D', u'1976Afz....12..665D', u'1983Ap.....19..134D', u'1983Afz....19..229D', u'1974ATsir.809....2D', u'1984Afz....20..525D', u'1974Afz....10..315D', u'1984Ap.....20..290D', u'1973ATsir.759....6D', u'1974ATsir.837....2D', u'1974ATsir.809....1D', u'1971ATsir.621....7D'],
+                    'itemCount': 0,
+                    }]
+        response = LinkRequest(bibcode='1971ATsir.615....4D', link_type='ASSOCIATED').request_link_type_associated(results)
+        self.assertEqual(response._status_code, 400)
+        self.assertEqual(response.response[0], '{"KeyError": "key not in dictionary"}')
+
+
+    def test_link_associated_IndexError(self):
+        """
+        return 404 from request_link_type_associated when there is IndexError
+        :return:
+        """
+        results = [{'bibcode': u'1971ATsir.615....4D',
+                    'link_type': u'ASSOCIATED',
+                    'link_sub_type': '',
+                    'title': [u'Part 11'],
+                    'url': [u'1971ATsir.615....4D', u'1971ATsir.624....1D', u'1976Afz....12..665D', u'1983Ap.....19..134D', u'1983Afz....19..229D', u'1974ATsir.809....2D', u'1984Afz....20..525D', u'1974Afz....10..315D', u'1984Ap.....20..290D', u'1973ATsir.759....6D', u'1974ATsir.837....2D', u'1974ATsir.809....1D', u'1971ATsir.621....7D'],
+                    'itemCount': 0,
+                    }]
+        response = LinkRequest(bibcode='1971ATsir.615....4D', link_type='ASSOCIATED').request_link_type_associated(results)
+        self.assertEqual(response._status_code, 400)
+        self.assertEqual(response.response[0], '{"IndexError": "list index out of range"}')
```

Comment:
i think the error message would be clearer saying something like: 'Requested link_type=%s is missing for the bibcode-%s'

Reply:
thank you so much. fixed.